File: epi/PowerXY.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class PowerXY {
  @EpiTest(testDataFile = "power_x_y.tsv")
  public static double power(double x, int y) {
//    int absY = Math.abs(y);
//
//    double rt = 1;
//    while(absY != 0) {
//      if((absY & 1) != 0) {
//        rt *= x;
//      }
//
//      x *= x;
//      absY >>>= 1;
//    }
//
//    return y < 0 ? 1 / rt : rt;
      return epi.kt.PowerXY.INSTANCE.power(x, y);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "PowerXY.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/PowerXY.kt
package epi.kt

import kotlin.math.abs

object PowerXY {
    fun power(x: Double, y: Int): Double {
        var absY = abs(y)
        var rt = 1.0
        var lX = x
        while(absY != 0) {
            if((absY and 1) != 0) {
                rt *= lX
            }

            lX *= lX
            absY = absY ushr(1)
        }

        return if(y < 0) {
            1 / rt
        } else {
            rt
        }
    }
}